### ISSUE TITLE
Has swift 1.2 solved your issues?

### ISSUE THREAD

**Issue by cognociente** [00buggy00/SwiftOpenGL#1]
Found your stack overflow post: http://stackoverflow.com/questions/25981553/cvdisplaylink-with-swift
where you mention that Swift 1.2 could remove the need for objective-c code... did this happen?

**Comment by cognociente** [comment]
My callback function looks like the following (Swift 1.2) and it seems to pass through the Swift compiler.

```
static func ScrollingCallback(displayLink:CVDisplayLinkRef, now: UnsafePointer<CVTimeStamp>, 
outputTime: UnsafePointer<CVTimeStamp>, flagsIn:CVOptionFlags,  flagsOut:CVOptionFlags, 
displayLinkContext: Unmanaged<AnyObject>)->CVReturn {

    var clipView:ClipView = displayLinkContext.takeRetainedValue() as! ClipView
    dispatch_async(dispatch_get_main_queue(), {clipView.updateOrigin})
    return 0
}
```

I don't know if the above is correct BUT what I can't figure out how to do is use it in the DisplayLink function setting the callback (which you seem to have figured out):

```
CVDisplayLinkSetOutputCallback(displayLink!, &ClipView.ScrollingCallback(), nil) //Compiler NOT happy
```

**Comment by BuggusMageevers** [comment]
Thank you for your comment.  I hope that you found my StackOverflow answer helpful as well.  It may be a typo, but in your comment, your flagsOut should by an `UnsafePointer<CVOptionFlags>`.  Also, for the `displayLinkContext`, I used an `UnsafeMutablePointer<Void>` when I was doing my testing, and you may certainly try that.  If you have and it was giving you compiler errors, then please please disregard.  

``` Swift
func ScrollingCallback(displayLink: CVDisplayLink , now: UnsafePointer<CVTimeStamp>, outputTime: UnsafePointer<CVTimeStamp>, flagsIn: CVOptionFlags, flagsOut: UnsafePointer<CVOptionFlags>, displayLinkContext: UnsafeMutablePointer<Void>)->CVReturn { }
```

Currently, I don't have the money to spend on a Developer membership.  I'm just a hobbiest programmer.  As such, I don't currently have access to Swift 1.2, I can just see what is being developed.  Concerning my  StackOverflow answer, I did not intend to imply that Swift 1.2 would potentially solve the issue of declaring a callback in Swift that could be used with one of the C API's (like CVDisplayLink) in Swift.  Please accept my apology if this is the case.  Swift 1.2, according the most recent release notes, has added the following C API support.

> ### New in Xcode 6.3 beta
> 
> #### Swift Language Enhancements
> 
> Swift can now partially import C aggregates containing unions, bitfields, SIMD vector types, and other C language features that are not natively supported in Swift.
> 
> Imported C structs now have a default initializer in Swift...

However, it does not add the support to pass a pointer to a Swift function to functions in a C API.  It's terribly unfortunate.  That being said, I hope that this functionality may appear in the near future.  Apple seems to be actively pursuing a purely Swift development environment--which makes sense.  If Apple were to announce this new language, but then decide to only partially incorporate it into their business, it would most likely die as a project and turn into one of Apple's major blunders.  

Swift's growing ability to pass around C data more easily to C functions and Obj-C methods is reassuring.  Over the next couple years, Swift will, hopefully have developed enough to take over entirely.  That's the great part about Swift, it's low level enough to push bits around, but high level enough to enable OOP and some aspects of functional programming.

Should you find a solution that you feel works better than using a bridge over into Obj-C/C, I would greatly appreciate your sharing such method.  I shall do the same in some future pushes to GITHub as I develop salient code samples that incorporate Swift and OpenGL.  Currently, I'm seeing how functional programming may be used to harness C-like API's more effectively with cleaner code.  If you have some insight, I would greatly appreciate it.

**Comment by BuggusMageevers** [comment]
Anyone awaiting the resolution of the issue, my apologies in my delinquency.  The issue regarding the callback function not being supportable with completely Swift code has been resolved as of Swift 3.0.  I've updated the current project here to something more similar to what I am currently using with Swift 4.x.
